### PR TITLE
Handle big integers in comparison and sprintf

### PR DESCRIPTION
**Pull request by andrehaland**
This fix makes it possible to compare big integers,
as well as casting big integers to string with builtin
`sprintf()` function.

YAML test cases for `sprintf()` were not added due to bug in
`go-yaml` when unmarshalling big integers
 (see https://github.com/open-policy-agent/opa/issues/3147#issuecomment-780660884 and https://github.com/open-policy-agent/opa/issues/3147#issuecomment-781114686). 
Issue has been raised: https://github.com/go-yaml/yaml/issues/703.
A standard go test file was added instead.

Fixes: #3147
Signed-off-by: Andre Håland <andre.haland@gmail.com>

### CODE REVIEW

**Comment by srenatus** [comment]
Thanks a lot for looking into this, providing a PR and **also filing the go-yaml issue** :+1:

While this looks great I think we've got a few things to think about or talk through with this change:

1. backwards compatibility: clearly, this is changing the outcome of a comparison of two big numbers: where before they were considered equal, they'll now not be.

    I suppose we can call it a bug fix, though? After all, the numbers' representation in rego looks different, but they were considered equal.
2. alternative approaches: I think it was @koponen-styra who brought up using `new(big.Float).SetPrec(big.MaxPrec).SetString(..)` instead of what we do right now.

   How do these approaches differ? I'd guess `big.Rat` is more accurate than `big.MaxPrec` -- but that'll come at a price, too, I guess?
3. string representations: What this was brought up, I think, we've talked about the need to use [`FloatString`](https://golang.org/pkg/math/big/#Rat.FloatString) to keep our string representations intact (we wouldn't want "7/4" to show up now, but "1.75").

    I guess we've got no test coverage for those, since this PR got green....? 🤔

**Comment by johanneslarsson** [comment]
> Thanks a lot for looking into this, providing a PR and **also filing the go-yaml issue** 👍
> 
> While this looks great I think we've got a few things to think about or talk through with this change:
> 
> 1. backwards compatibility: clearly, this is changing the outcome of a comparison of two big numbers: where before they were considered equal, they'll now not be.
>    I suppose we can call it a bug fix, though? After all, the numbers' representation in rego looks different, but they were considered equal.
> 2. alternative approaches: I think it was @koponen-styra who brought up using `new(big.Float).SetPrec(big.MaxPrec).SetString(..)` instead of what we do right now.
>    How do these approaches differ? I'd guess `big.Rat` is more accurate than `big.MaxPrec` -- but that'll come at a price, too, I guess?
> 3. string representations: What this was brought up, I think, we've talked about the need to use [`FloatString`](https://golang.org/pkg/math/big/#Rat.FloatString) to keep our string representations intact (we wouldn't want "7/4" to show up now, but "1.75").
>    I guess we've got no test coverage for those, since this PR got green....? 🤔

1. My guess is that comparing as large numbers as these are unusual. But we can't be sure. The reason it become so big in this case is that serial numbers are randomised today instead of sequential as it used to be (pre 2016). I doubt any CA issued more certs than the max.int during its lifetime.
2. Probably worth a look as well, but how would one know which solution that is most performant? Do `go perf` give us an indication here or do we need additional tests?
3. Rat isn't part of the sprintf function. As its not possible to use the testrunner for this, I guess its applicable to add a topdown/strings_test.go ?

**Comment by andrehaland** [comment]
> 2\. alternative approaches: I think it was @koponen-styra who brought up using `new(big.Float).SetPrec(big.MaxPrec).SetString(..)` instead of what we do right now.
>     How do these approaches differ? I'd guess `big.Rat` is more accurate than `big.MaxPrec` -- but that'll come at a price, too, I guess?

As for the precision, I see that there currently is a restriction in the parser on the size of numbers due to non-linear scaling of printing big numbers:
https://github.com/open-policy-agent/opa/blob/930d1031b94ea06a3bdfb7b3020829e1745c23a1/ast/parser.go#L886-L897

**Comment by srenatus** [comment]
> As for the precision, I see that there currently is a limitation in the parser on the size of numbers due to non-linear scaling of printing big numbers [...]

That would mean the proper value to pass to `SetPrec` is something matching what the parser does, but less than `big.MaxPrec`...? Or would there be no penalty from setting it to MaxPrec while the parser actually restricts these some more...? 🤔 

From some internal discussion, I think there's strong support that we keep using `big.Float` instead of `big.Rat`. -- Sorry for shoving you on that track... 💦

**Comment by andrehaland** [comment]
I ran a performance test comparing `big.Float` (current implementation at master branch) and `big.Float.SetPrec(big.MaxPrec)` with `go test -tags=slow -timeout=30m -run=- -bench=. -benchmem -count 10 ./...`. Even with a timeout of 30m, the `test/scheduler` test with SetPrec(big.MaxPrec) failed due to timeout, indicating that there are some problems with setting such high precision. I am looking into this.

```
BenchmarkScheduler10x30-4   	SIGQUIT: quit
PC=0x5eb122 m=0 sigcode=0

goroutine 73 [running]:
math/big.divWVW(0xc0605bc000, 0x4000001, 0x4000005, 0x0, 0xc0405ba000, 0x4000001, 0x4000001, 0xad78ebc5ac620000, 0x467f60, 0x4000001, ...)

...

*** Test killed with quit: ran too long (31m0s).
exit status 2
```
Output where "old" is current `big.Float` and "new" is `big.Float.SetPrec(big.MaxPrec)`:
[benchmark.txt](https://github.com/open-policy-agent/opa/files/6041382/benchmark.txt)

**Comment by andrehaland** [comment]
Looking at this playground https://play.golang.org/p/Nt7cJUZEg_X, it seems that more memory is allocated as we increase the precision. It also seems that printing `big.Float`'s with high precision takes alot of time (playground timeouts on the one with precision 2e5).

**Comment by andrehaland** [comment]
So it seems that the Go playground has some timeout settings, so I am not able to test for higher precisions there. Running this Go program on my local computer:
```go
package main

import (
	"fmt"
	"math/big"
	"time"
)

func main(){
	f:= createBigFloatFromString("1.123456789123456789123", 1e5)
	printBigFloat(f)
}


func createBigFloatFromString(s string, prec uint) *big.Float{
	defer timeTrack(time.Now(), "create")
	f, ok := new(big.Float).SetPrec(prec).SetString(s)
	if !ok {
		panic(ok)
	}
	return f
}

func printBigFloat(f *big.Float) {
	defer timeTrack(time.Now(), "print")
	fmt.Printf("print: %v\n", f)
}

func timeTrack(start time.Time, name string) {
	elapsed := time.Since(start)
	fmt.Printf("%s took %s\n", name, elapsed)
}
```
gives:
```
go build . && ./bigfloats
create took 105.978µs
print: 1.123456789123456789123
print took 642.096185ms
```

**Comment by koponen-styra** [comment]
> Looking at this playground https://play.golang.org/p/Nt7cJUZEg_X, it seems that more memory is allocated as we increase the precision. It also seems that printing `big.Float`'s with high precision takes alot of time (playground timeouts on the one with precision 2e5).

Interesting. Looking at the big.Float internals, it indeed looks like big.Float proactively allocates per the configured precision. big.Rat.SetString() seems to avoid this. That in mind, big.Rat based approach for *comparing* seems justified but it sure would be to good have a comment or two around this; using big.Rat to parse a floating point for maximal accuracy to allow for minimal memory allocation sure is not obvious at all...!

File: topdown/strings_test.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package topdown
+
+import (
+	"context"
+	"testing"
+
+	"github.com/open-policy-agent/opa/ast"
+)
+
+func TestSprintf(t *testing.T) {
```
Comment (johanneslarsson) [inline topdown/strings_test.go:10]: Perhaps rename this function to TestSprintfBigInt and also add a comment why this isn't possible to run from the test runner (yet)?

**Review by srenatus — DISMISSED** [review]
I think we should go ahead with this. Thanks for all your work on this 🎉 🚀

**Review by srenatus — APPROVED** [review]
LGTM, thanks again!

Follow-up stuff I've made a note of:
- checking wasm, by...
- trying to encode the test case in the yaml, as part of the module (having the query only query for the rule, not parse any numbers)

**Pull request merged**